File: design-add-and-search-words-data-structure/flynn.cpp
/**
 * 풀이
 * - Trie 구조를 활용하여 풀이할 수 있습니다
 * - wildcard인 '.'에 대한 처리가 필요합니다
 * 
 * Big O
 * - N: 주어지는 문자열 word의 길이
 * - M: 현재 WordDictionary에 저장되어 있는 TrieNode의 수
 * 
 * - void addWord(string word)
 *   - Time complexity: O(N)
 *   - Space complexity: O(N)
 *     - 최악의 경우 word의 모든 문자에 대해 새로운 TrieNode를 추가해야 합니다
 * 
 * - bool search(string word)
 * - bool _search(string word, int idx, TrieNode* node)
 *   - Time complexity: best O(N), worst O(M) < O(26^N)
 *     - wildcard 사용 및 기존에 저장된 word의 상태에 따라 달라집니다
 *   - Space complexity: O(N)
 *     - _search가 재귀적으로 호출되므로 재귀 호출 스택의 깊이만큼 추가적인 공간이 사용됩니다
 *     - 재귀 호출 스택의 깊이는 현재 찾는 word의 길이에 선형적으로 비례합니다
 */

class TrieNode {
public:
    array<TrieNode*, 27> links;
    bool word;

    TrieNode(): word(false) {
        links.fill(nullptr);
    }
};

class WordDictionary {
public:
    WordDictionary(): root(new TrieNode()) {}
    
    void addWord(string word) {
        TrieNode* current = root;
        
        for (char c : word) {
            if (current->links[c - 'a'] == nullptr) {
                current->links[c - 'a'] = new TrieNode();
            }
            current = current->links[c - 'a'];
        }

        current->word = true;
    }
    
    bool search(string word) {
        return _search(word, 0, root);
    }
    
private:
    TrieNode* root;

    bool _search(string word, int idx, TrieNode* node) {
        if (word.size() == idx) return node->word;

        char c = word[idx];

        if (c != '.') {
            if (node->links[c - 'a'] == nullptr) return false;
            
            TrieNode* next_node = node->links[c - 'a'];
            int next_idx = idx + 1;

            return _search(word, next_idx, next_node);
        } else {
            for (TrieNode* link : node->links) {
                if (link != nullptr) {
                    TrieNode* next_node = link;
                    int next_idx = idx + 1;

                    if (_search(word, next_idx, next_node)) return true;
                }
            }
            return false;
        }
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: word-search-ii/flynn.go
/*
풀이
- Trie 자료구조와 2차원 배열에서의 DFS & Backtracking을 사용하여 풀이할 수 있습니다
Big O
- M, N: 주어진 2차원 배열 board의 행, 열 크기
- Max(W): 주어진 배열 words 중에서 가장 길이가 긴 단어의 길이
- Sum(W): 주어진 배열 words에 포함된 모든 단어의 길이의 총합
- Time complexity: O(Sum(W) + MN * Max(W))
  - building trie: O(Sum(W))
    - 각 단어의 모든 문자를 한 번씩 조회합니다
  - 반복문: O(MN * Max(W))
    - 행, 열에 대한 반복: O(MN)
    - dfs: O(Max(W))
      - 한 좌표에 대해 dfs 함수를 trie의 최대 깊이, 즉 O(Max(W))만큼 호출합니다
- Space complexity: O(Sum(W) + Max(W))
    - building trie: O(Sum(W)) at worst
    - dfs: O(Max(W))
      - 재귀 호출 스택의 깊이를 고려해야 합니다
*/

func findWords(board [][]byte, words []string) []string {
	// building trie
	root := &trieNode{}
	for _, w := range words {
		root.add(w)
	}

	m := len(board)
	n := len(board[0])
	res := make([]string, 0, len(words)) // create result
	for r := 0; r < m; r++ {
		for c := 0; c < n; c++ {
			if len(res) == len(words) { // early break if we found all the words
				break
			}
			if root.children[idx(board[r][c])] != nil { // dfs if you found starting letter of any words
				dfs(r, c, &board, root, &res)
			}
		}
	}

	return res
}

// ----- Implementation of TrieNode -----
type trieNode struct {
	children [26]*trieNode
	word     string
}

func (t *trieNode) add(w string) {
	for i, c := range w {
		if t.children[c-'a'] == nil { // create new child node if there wasn't
			t.children[c-'a'] = &trieNode{}
		}
		t = t.children[c-'a']
		if i == len(w)-1 {
			t.word = w
		}
	}
}

// ----- Dfs -----
func dfs(r, c int, board *[][]byte, parentNode *trieNode, res *[]string) {
	currLetter := (*board)[r][c]
	currNode := parentNode.children[idx(currLetter)]
	// if current trie node represents some word, then append it to the result
	if currNode.word != "" {
		*res = append(*res, currNode.word)
		currNode.word = "" // prevent duplicate words into the result
	}
	// mark current cell as visited
	(*board)[r][c] = '#'
	// search for the 4 directions
	m := len(*board)
	n := len((*board)[0])
	dr := []int{0, 0, 1, -1}
	dc := []int{1, -1, 0, 0}
	for i := 0; i < 4; i++ {
		nr := r + dr[i]
		nc := c + dc[i]
		if !(0 <= nr && nr < m) || !(0 <= nc && nc < n) { // skip the invalid coordinates
			continue
		}
		if (*board)[nr][nc] == '#' { // skip the visited cell
			continue
		}
		if currNode.children[idx((*board)[nr][nc])] != nil {
			dfs(nr, nc, board, currNode, res)
		}
	}
	// restore the cell
	(*board)[r][c] = currLetter
}

// ----- Helper -----
func idx(b byte) int {
	return int(b - 'a')
}
